kernel/src/syscall/timerfd.rs

```diff
@@ -0,0 +1,257 @@
+    fd: FileDesc,
+    flags: i32,
+    new_itimerspec_addr: Vaddr,
+    old_itimerspec_addr: Vaddr,
+    ctx: &Context,
+) -> Result<SyscallReturn> {
+    let file_table = ctx.thread_local.borrow_file_table();
+    let file_table_locked = file_table.unwrap().read();
+    let timerfd_file = file_table_locked.get_file(fd as _)?;
+    let timerfd_file = timerfd_file
+        .downcast_ref::<TimerfdFile>()
+        .ok_or_else(|| Error::with_message(Errno::EINVAL, "the fd is not a timerfd"))?;
+
+    let user_space = ctx.user_space();
+    let new_itimerspec = user_space.read_val::<itimerspec_t>(new_itimerspec_addr)?;
+    let interval = Duration::try_from(new_itimerspec.it_interval)?;
+    let expire_time = Duration::try_from(new_itimerspec.it_value)?;
+
+    let timer = &timerfd_file.timer;
+    if old_itimerspec_addr > 0 {
+        let old_interval = timespec_t::from(timer.interval());
+        let remain = timespec_t::from(timer.remain());
+        let old_itimerspec = itimerspec_t {
+            it_interval: old_interval,
+            it_value: remain,
+        };
+        user_space.write_val(old_itimerspec_addr, &old_itimerspec)?;
+    }
+
+    timer.set_interval(interval);
+    timer.cancel();
+    // Clear `ticks` after cancel the timer to ensure that `ticks` is zero
+    // when the timer is rearmed.
+    timerfd_file.ticks.store(0, Ordering::Release);
+
+    const TFD_TIMER_CANCEL_ON_SET: i32 = 2;
+    if expire_time != Duration::ZERO {
+        let timeout = if flags == 0 {
+            Timeout::After(expire_time)
+        } else if flags == TIMER_ABSTIME {
+            Timeout::When(expire_time)
+        } else if flags == TFD_TIMER_CANCEL_ON_SET {
+            unimplemented!()
+        } else {
+            unreachable!()
+        };
+        timer.set_timeout(timeout);
+    }
+
+    Ok(SyscallReturn::Return(0))
+}
+
+pub fn sys_timerfd_gettime(
+    fd: FileDesc,
+    itimerspec_addr: Vaddr,
+    ctx: &Context,
+) -> Result<SyscallReturn> {
+    if itimerspec_addr == 0 {
+        return_errno_with_message!(Errno::EINVAL, "invalid pointer to return value");
+    }
+
+    let file_table = ctx.thread_local.borrow_file_table();
+    let file_table_locked = file_table.unwrap().read();
+    let timerfd_file = file_table_locked.get_file(fd as _)?;
+    let timerfd_file = timerfd_file
+        .downcast_ref::<TimerfdFile>()
+        .ok_or_else(|| Error::with_message(Errno::EINVAL, "the fd is not a timerfd"))?;
+
+    let interval = timespec_t::from(timerfd_file.timer.interval());
+    let remain = timespec_t::from(timerfd_file.timer.remain());
+    let itimerspec = itimerspec_t {
+        it_interval: interval,
+        it_value: remain,
+    };
+    ctx.user_space().write_val(itimerspec_addr, &itimerspec)?;
+
+    Ok(SyscallReturn::Return(0))
+}
+
+bitflags! {
+    struct Flags: u32 {
+        const TFD_CLOEXEC = CreationFlags::O_CLOEXEC.bits();
+        const TFD_NONBLOCK = StatusFlags::O_NONBLOCK.bits();
+    }
+}
+
+struct TimerfdFile {
```

Comment:
In the system call layer, each system call is considered to be significant enough to have a dedicated file, even when such a dedicated file would be small. To follow this convention, first move `TimerfdFile` out of the system call layer, probably into the `kernel/src/time` and then extract each system call function into a separate file.